include/cantera/kinetics/Arrhenius.h

```diff
@@ -468,7 +381,73 @@ class BlowersMaselRate final : public ArrheniusBase, public ReactionRate
     double m_deltaH_R; //!< enthalpy change of reaction (in temperature units)
 };
 
-}
 
+//! A class template for bulk phase reaction rate specifications
+template <class RateType, class DataType>
+class BulkRate final : public RateType
+{
+public:
+    BulkRate() = default;
+    using RateType::RateType; // inherit constructors
+
+    //! Constructor based on AnyMap content
+    BulkRate(const AnyMap& node, const UnitStack& rate_units={}) {
+        setParameters(node, rate_units);
+    }
+
+    unique_ptr<MultiRateBase> newMultiRate() const override {
+        return unique_ptr<MultiRateBase>(
+            new MultiRate<BulkRate<RateType, DataType>, DataType>);
+    }
+
+    virtual void setParameters(
+        const AnyMap& node, const UnitStack& rate_units) override
+    {
+        RateType::m_negativeA_ok = node.getBool("negative-A", false);
+        if (!node.hasKey("rate-constant")) {
+            RateType::setRateParameters(AnyValue(), node.units(), rate_units);
+            return;
+        }
+        RateType::setRateParameters(node["rate-constant"], node.units(), rate_units);
+    }
+
+    virtual void getParameters(AnyMap& node) const override {
+        if (RateType::m_negativeA_ok) {
+            node["negative-A"] = true;
+        }
+        AnyMap rateNode;
+        RateType::getRateParameters(rateNode);
+        if (!rateNode.empty()) {
+            // RateType object is configured
+            node["rate-constant"] = std::move(rateNode);
+        }
+        if (RateType::type() != "Arrhenius") {
+            node["type"] = RateType::type();
+        }
+    }
+
+    //! Evaluate reaction rate
+    /*!
+     *  @param shared_data  data shared by all reactions of a given type
+     */
+    double evalFromStruct(const DataType& shared_data) const {
+        return RateType::evalRate(shared_data);
+    }
```

Comment:
If you rename `evalRate(DataType&)` to `evalFromStruct(DataType&)` in `ArrheniusBase` and friends, I think this method can just be deleted. Likewise for `ddTScaledFromStruct` below.

Reply:
Correct, although they would have to be declared virtual, as there will be different implementations in other anticipated class templates. Due to the `final` declaration it should not make any difference. At the same time, the compiler inlines this and it does not appear to make any difference (I already tested this).

Reply:
*Sorry, I had clicked the wrong button -- meant to add this as part of a review.*

It's perfectly acceptable to override a non-virtual method in a derived class. In such a case, calls to this function will always be based on the visible type of the object, not a more derived type. This can be (and is) resolved at compile time, which is exactly what we want in this case, since the `MultiRate` class always has a pointer to the most-derived type, and no virtual function calls are needed (which furthermore means there's no need to mark the class as `final`).